Libraries/LibWeb/HTML/HTMLOptionElement.cpp

```diff
@@ -194,27 +202,73 @@ bool HTMLOptionElement::disabled() const
+    // The form getter steps are:
+
+    // 1. Let select be this's option element nearest ancestor select.
+    auto select = nearest_select_element();
+
+    // 2. If select is null, then return null.
+    if (!select)
+        return {};
+
+    // 3. Return select's form owner.
+    return select->form();
 }
 
-GC::Ptr<HTMLSelectElement> HTMLOptionElement::owner_select_element()
+// https://html.spec.whatwg.org/multipage/form-elements.html#update-an-option's-nearest-ancestor-select
+void HTMLOptionElement::update_nearest_select_element()
 {
-    if (auto* maybe_parent = parent()) {
-        if (auto* select_element = as_if<HTMLSelectElement>(*maybe_parent))
-            return select_element;
-        if (auto* opt_group_element = as_if<HTMLOptGroupElement>(*maybe_parent)) {
-            if (auto* maybe_parent = opt_group_element->parent())
-                return as_if<HTMLSelectElement>(*maybe_parent);
+    // 1. Let oldSelect be option's cached nearest ancestor select element.
+    auto old_select = m_cached_nearest_select_element;
+
+    // 2. Let newSelect be option's option element nearest ancestor select.
+    auto new_select = compute_nearest_select_element();
+
+    // 3. If oldSelect is not newSelect:
+    if (old_select != new_select) {
+        // 1. If oldSelect is not null, then run the selectedness setting algorithm given oldSelect.
+        if (old_select)
+            old_select->update_selectedness();
+
+        // 2. If newSelect is not null, then run the selectedness setting algorithm given newSelect.
+        if (new_select)
+            new_select->update_selectedness();
+    }
+
+    // 4. Set option's cached nearest ancestor select element to newSelect.
+    m_cached_nearest_select_element = new_select;
+}
+
+// https://html.spec.whatwg.org/multipage/form-elements.html#option-element-nearest-ancestor-select
+GC::Ptr<HTMLSelectElement> HTMLOptionElement::compute_nearest_select_element()
+{
+    // 1. Let ancestorOptgroup be null.
+    GC::Ptr<HTMLOptGroupElement> ancestor_optgroup;
+
+    // 2. For each ancestor of option's ancestors, in reverse tree order:
+    for (auto* ancestor = parent(); ancestor; ancestor = ancestor->parent()) {
+        // 1. If ancestor is a datalist, hr, or option element, then return null.
+        if (is<HTMLDataListElement>(*ancestor)
+            || is<HTMLHRElement>(*ancestor)
+            || is<HTMLOptionElement>(*ancestor))
+            return nullptr;
+
+        // 2. If ancestor is an optgroup element:
+        if (is<HTMLOptGroupElement>(*ancestor)) {
+            // 1. If ancestorOptgroup is not null, then return null.
+            if (ancestor_optgroup)
+                return nullptr;
+
+            // 2. Set ancestorOptgroup to ancestor.
+            ancestor_optgroup = as<HTMLOptGroupElement>(*ancestor);
         }
+
+        // 3. If ancestor is a select, then return ancestor.
+        if (is<HTMLSelectElement>(*ancestor))
+            return as<HTMLSelectElement>(*ancestor);
```

Comment:
For cases where you want to both check a type and also cast to it, use `as_if<>()`, which avoids doing type checking twice. eg:

```c++
if (auto* select_element = as_if<HTMLSelectElement(*ancestor))
	return select_element;
```

Same goes for step 2.